docs/concepts/unions.md

```diff
@@ -76,25 +76,59 @@ print(User(id='456'))  # (2)
 2. We're in lax mode and the numeric string `'123'` is valid as input to the first member of the union, `int`.
    Since that is tried first, we get the surprising result of `id` being an `int` instead of a `str`.
 
+!!! tip
+    For complex cases, it's recommended to use either [`left_to_right`](#left-to-right-mode) mode or [discriminated unions](#discriminated-unions).
+    With these approaches, you can control which member of the union to validate against, and in which order.
+    If you're looking for incredibly specialized behavior, you can use a [custom validator](../concepts/validators.md#field-validators).
+
 ### Smart Mode
 
 Because of the surprising side effects of `union_mode='left_to_right'`, in Pydantic >=2 the default mode for `Union` validation is `union_mode='smart'`.
 
-In this mode, pydantic scores a match of a union member into one of the following three groups (from highest score to lowest score):
+In this mode, pydantic selects the best match for the input from the union members based on two main factors:
+
+1. The number of valid fields set (relevant for models, dataclasses, and typed dicts)
```

Comment:
I'd be then tempted to put the actual algorithm (i.e. everything from here down) behind a `<details>` collapsible tag so that interested parties can read it, while subtly emphasising further that the algorithm is a mostly opaque and internal concern which may change.